### PR TITLE
Support for Requester Pays

### PR DESCRIPTION
**Pull request by HaileyJang**
Skyplane CLI will now take in additional flag to support requester pays bucket. Requester-pays bucket are buckets that charges the person who downloads the objects. More information in the following link: https://docs.aws.amazon.com/AmazonS3/latest/userguide/RequesterPaysBuckets.html

### CODE REVIEW

File: skyplane/chunk.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ class ChunkRequest:
     src_random_size_mb: Optional[int] = None
     src_object_store_bucket: Optional[str] = None
     dst_object_store_bucket: Optional[str] = None
+    src_requester_pays: str = ''
```
Comment (parasj) [inline skyplane/chunk.py:50]: We should move this to the gateway startup code as a flag

File: skyplane/cli/cli.py [hunk 2]
```diff
@@ -65,6 +66,7 @@ def cp(
         skyplane_root / "profiles" / "throughput.csv", "--throughput-grid", help="Throughput grid file"
     ),
     solver_verbose: bool = False,
+    requester_pays: str = typer.Option(None, "--requester-pays", help = "Bucket flag for Requester pays buckets")
```
Comment (parasj) [inline skyplane/cli/cli.py:69]: We should make this a config flag in config.py and make this a boolean flag

File: skyplane/cli/cli_impl/cp_replicate.py [hunk 3]
```diff
@@ -127,22 +127,25 @@ def generate_full_transferobjlist(
...
     dest_prefix: str,
     recursive: bool = False,
+    requester_pays: str = None
 ) -> List[Tuple[ObjectStoreObject, ObjectStoreObject]]:
     """Query source region and destination region buckets and return list of objects to transfer."""
     source_iface = ObjectStoreInterface.create(source_region, source_bucket)
     dest_iface = ObjectStoreInterface.create(dest_region, dest_bucket)
 
+    if requester_pays == 'requester' and isinstance(source_iface, S3Interface):
+        source_iface.requester_pays = True
```
Comment (parasj) [inline skyplane/cli/cli_impl/cp_replicate.py:137]: Actually, add a method like `set_requester_pays()` to `ObjectStoreInterface` with abstract methods (silently do nothing for GCP and Azure).

File: skyplane/cli/cli_impl/cp_replicate.py [hunk 4]
```diff
@@ -127,22 +127,25 @@ def generate_full_transferobjlist(
...
         raise exceptions.MissingBucketException(f"Source bucket {source_bucket} does not exist")
     if not dest_iface.bucket_exists():
         raise exceptions.MissingBucketException(f"Destination bucket {dest_bucket} does not exist")
-
     source_objs, dest_objs = [], []
 
     # query all source region objects
     logger.fs.debug(f"Querying objects in {source_bucket}")
-    with console.status(f"Querying objects in {source_bucket}") as status:
+    with console.status(f"Querying objects in {source_bucket}\n") as status:
```
Comment (parasj) [inline skyplane/cli/cli_impl/cp_replicate.py:148]: Print the new line with the logger statement `logger.error("Specified object does not exist.")`

File: skyplane/cli/cli_impl/cp_replicate.py [hunk 5]
```diff
@@ -233,6 +236,7 @@ def launch_replication_job(
     use_compression: bool = False,
     use_e2ee: bool = True,
     use_socket_tls: bool = False,
+    requester_pays: str = '',
```
Comment (parasj) [inline skyplane/cli/cli_impl/cp_replicate.py:239]: Use a bool for this

File: skyplane/obj_store/s3_interface.py [hunk 6]
```diff
@@ -46,7 +48,14 @@ def _s3_client(self, region=None):
 
     def bucket_exists(self):
         s3_client = self._s3_client("us-east-1")
-        return self.bucket_name in [b["Name"] for b in s3_client.list_buckets()["Buckets"]]
+        if self.requester_pays:
+            try:
```
Comment (parasj) [inline skyplane/obj_store/s3_interface.py:52]: Do this for everything, not just requester pays

Comment (parasj) [reply skyplane/obj_store/s3_interface.py:52]: Add a limit = 1 for this call

File: skyplane/obj_store/s3_interface.py [hunk 7]
```diff
@@ -63,8 +72,15 @@ def delete_bucket(self):
     def list_objects(self, prefix="") -> Iterator[S3Object]:
         paginator = self._s3_client().get_paginator("list_objects_v2")
         page_iterator = paginator.paginate(Bucket=self.bucket_name, Prefix=prefix)
-        for page in page_iterator:
-            for obj in page.get("Contents", []):
+        if not self.requester_pays:
```
Comment (parasj) [inline skyplane/obj_store/s3_interface.py:75]: You should be able to set the RequesterPays using this method: https://stackoverflow.com/a/44376294

**Review by parasj — CHANGES_REQUESTED** [review]
Great work, this will really help our users! I added some minor comments to improve maintainability with a single code path for both workflows.

File: skyplane/obj_store/object_store_interface.py [hunk 8]
```diff
@@ -27,6 +27,9 @@ def path(self) -> str:
     def region_tag(self) -> str:
         raise NotImplementedError()
 
+    def activate_requester(self):
```
Comment (parasj) [inline skyplane/obj_store/object_store_interface.py:30]: `activate_requester ` => `set_requester_pays_mode(status: bool)` so we can disable it as well

File: skyplane/obj_store/s3_interface.py [hunk 9]
```diff
@@ -43,14 +45,20 @@ def aws_region(self):
...
     def _s3_client(self, region=None):
         region = region if region is not None else self.aws_region
         return self.auth.get_boto3_client("s3", region)
 
     def bucket_exists(self):
         s3_client = self._s3_client("us-east-1")
         try:
-            s3_client.list_objects_v2(Bucket=self.bucket_name, MaxKeys=1)  # list one object to check if bucket exists
+            if self.requester_pays:
+                s3_client.list_objects(Bucket=self.bucket_name, RequestPayer="requester", MaxKeys=1)
```
Comment (parasj) [inline skyplane/obj_store/s3_interface.py:59]: Merge these cases as:
```python
requester_pays = {"RequestPayer": "requester"} if self.requester_pays else {}
s3_client.list_objects(Bucket=self.bucket_name, MaxKeys=1, **requester_pays)
```

File: skyplane/obj_store/s3_interface.py [hunk 10]
```diff
@@ -70,7 +78,10 @@ def delete_bucket(self):
 
     def list_objects(self, prefix="") -> Iterator[S3Object]:
         paginator = self._s3_client().get_paginator("list_objects_v2")
-        page_iterator = paginator.paginate(Bucket=self.bucket_name, Prefix=prefix)
+        if self.requester_pays:
```
Comment (parasj) [inline skyplane/obj_store/s3_interface.py:81]: Merge as above

File: skyplane/obj_store/s3_interface.py (hunk 9 above)
Comment (parasj) [reply skyplane/obj_store/s3_interface.py:59]: This will let us add other args for S3 options in the future as well

**Review by parasj — APPROVED** [review]
This looks good. Also update the error message here to note that the user may need to enable requester pays:
https://github.com/skyplane-project/skyplane/blob/0d875bd3cbf6a32345eca3624af30c870d46c8ab/skyplane/exceptions.py#L7-L11

`Please ensure that the bucket exists and is accessible. See https://skyplane.org/en/latest/faq.html#i-am-unable-to-copy-files-to-a-folder-in-my-destination-bucket-due-to-a-missing-object-error.```
For that documentation page, you should add more text to elaborate on requester pays setup.

**Comment by HaileyJang** [comment]
FAQ changes made in #538

**Review by parasj — APPROVED** [review]

**Pull request merged**